Remove hyper fork

DarrenTsung opened the pull request:
Use hyper 0.11 and associated libraries like tokio.

The flow of the project is largely unchanged to avoid having to refactor heavily downstream, but now we handle the threads ourselves with a RoundRobinPool and also use `Executor`s to receive messages via `futures::sync::mspc` from the pool thread.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/onesignal/hyper-client-pool/2)
<!-- Reviewable:end -->

DarrenTsung commented on src/deliverable.rs:
```diff
@@ -0,0 +1,14 @@
+
+impl Deliverable for mpsc::Sender<DeliveryResult> {
+    fn complete(self, result: DeliveryResult) {
+        let _ = self.send(result);
```
Good point, yeah it's unimportant because it's only used for the tests. I moved it over to the test mod, we don't really need to expose the `Deliverable` impl for `mpsc::Sender` anyways..

jwilm commented:
Reviewed 5 of 10 files at r1, 1 of 2 files at r2, 3 of 3 files at r3, 1 of 5 files at r4.
Review status: 5 of 9 files reviewed at latest revision, 1 unresolved discussion.

---

*[src/config.rs, line 18 at r3](https://reviewable.io/reviews/onesignal/hyper-client-pool/2#-LCZwl0P0W9YGTPsrbli:-LCZwl0P0W9YGTPsrblj:b1gitb2) ([raw file](https://github.com/onesignal/hyper-client-pool/blob/ce241e14d1bfbd5552429bc3383d4067ba17f1fb/src/config.rs#L18)):*
> ```Rust
> 
>     /// Number of DNS threads per worker
>     pub dns_threads_per_worker: usize,
> ```

Does hyper 0.11 still handle DNS? Seems like we should be able to plugin a tokio/futures-compatible resolver library and eliminate this option.

---

*[src/error.rs, line 9 at r3](https://reviewable.io/reviews/onesignal/hyper-client-pool/2#-LCZy8JoA6GqlYVPjwBW:-LCZy8JoA6GqlYVPjwBX:bwpv2vi) ([raw file](https://github.com/onesignal/hyper-client-pool/blob/ce241e14d1bfbd5552429bc3383d4067ba17f1fb/src/error.rs#L9)):*
> ```Rust
> 
> #[derive(Debug)]
> pub enum Error<D: Deliverable> {
> ```

The variants here need some explanation in doc comments. It would also be nice to expose a library `Error` type that buckets these into logical categories or something that is suggestive about how a consumer of this API should handle the errors. 

---

*[src/error.rs, line 12 at r3](https://reviewable.io/reviews/onesignal/hyper-client-pool/2#-LCZy0ba-p3pepCmDKDm:-LCZy0ba-p3pepCmDKDn:b-2hrafx) ([raw file](https://github.com/onesignal/hyper-client-pool/blob/ce241e14d1bfbd5552429bc3383d4067ba17f1fb/src/error.rs#L12)):*
> ```Rust
>     ThreadSpawn(io::Error),
>     HttpsConnector(hyper_tls::Error),
>     Full(Transaction<D>),
> ```

`Full`? What does that mean?

---

*[src/error.rs, line 33 at r3](https://reviewable.io/reviews/onesignal/hyper-client-pool/2#-LCZygOc7bXbyMV44ibC:-LCZygOc7bXbyMV44ibD:b-3w47cn) ([raw file](https://github.com/onesignal/hyper-client-pool/blob/ce241e14d1bfbd5552429bc3383d4067ba17f1fb/src/error.rs#L33)):*
> ```Rust
>             Error::ThreadSpawn(err) => SpawnError::ThreadSpawn(err),
>             Error::HttpsConnector(err) => SpawnError::HttpsConnector(err),
>             _ => unreachable!(),
> ```

This doesn't seem unreachable

---

*[src/error.rs, line 43 at r3](https://reviewable.io/reviews/onesignal/hyper-client-pool/2#-LCZyeVVDBMdhej5jv8k:-LCZyeVVDBMdhej5jv8l:b3ydu8h) ([raw file](https://github.com/onesignal/hyper-client-pool/blob/ce241e14d1bfbd5552429bc3383d4067ba17f1fb/src/error.rs#L43)):*
> ```Rust
>             Error::Full(err) => RequestError::Full(err),
>             Error::FailedSend(err) => RequestError::FailedSend(err),
>             _ => unreachable!(),
> ```

Nor does this

---

*[src/executor.rs, line 17 at r3](https://reviewable.io/reviews/onesignal/hyper-client-pool/2#-LC_8Rz931iWUUfj9NJZ:-LC_8Rz931iWUUfj9NJ_:b-b93mcr) ([raw file](https://github.com/onesignal/hyper-client-pool/blob/ce241e14d1bfbd5552429bc3383d4067ba17f1fb/src/executor.rs#L17)):*
> ```Rust
> 
> use config::Config;
> use counter::{Counter, WeakCounter};
> ```

Reviewable says `counter` module was removed in latest revision...

---

*[src/executor.rs, line 40 at r3](https://reviewable.io/reviews/onesignal/hyper-client-pool/2#-LC_8Yzc3A3CL6SnFFnh:-LC_8Yzc3A3CL6SnFFni:b-csr0yx) ([raw file](https://github.com/onesignal/hyper-client-pool/blob/ce241e14d1bfbd5552429bc3383d4067ba17f1fb/src/executor.rs#L40)):*
> ```Rust
> 
> enum ExecutorState<D: Deliverable> {
>     Running(FuturesMpsc::UnboundedReceiver<ExecutorMessage<D>>),
> ```

We should use a bounded channel here. The `Sender` for a bounded channel includes a `try_send` method which allows attempting a non-blocking send.

---

*[src/executor.rs, line 89 at r3](https://reviewable.io/reviews/onesignal/hyper-client-pool/2#-LC_A1Ee2C86pKGSAdxz:-LC_A1Ee2C86pKGSAdy-:b-k6xj54) ([raw file](https://github.com/onesignal/hyper-client-pool/blob/ce241e14d1bfbd5552429bc3383d4067ba17f1fb/src/executor.rs#L89)):*
> ```Rust
> 
>         info!("Spawning Executor.");
>         let tls = TlsConnector::builder().and_then(|builder| builder.build()).map_err(SpawnError::HttpsConnector)?;
> ```

I find this whole `TlsConnector` thing pretty strange. We can't establish our own sockets?

How is hyper doing DNS? We **must not** use the built-in glibc `getaddrinfo` that it previously used via `ToSocketAddrs`. This is why we switched to c-ares at some point. We should switch the resolver to TrustDNS resolver if it's not already using that.

I suppose we could actually implement our own `TlsConnector` and not have to fork Hyper :)

---

*[src/executor.rs, line 159 at r3](https://reviewable.io/reviews/onesignal/hyper-client-pool/2#-LC_C2yZ7ll7JscV9HHM:-LC_C2yZ7ll7JscV9HHN:bvhtyt6) ([raw file](https://github.com/onesignal/hyper-client-pool/blob/ce241e14d1bfbd5552429bc3383d4067ba17f1fb/src/executor.rs#L159)):*
> ```Rust
>                                         );
>                                     },
>                                     ExecutorMessage::Shutdown => {
> ```

I'm not a big fan of the `Shutdown` message for terminating an executor. Dropping all of the send handles guarantees no messages arrive after detecting the channel has closed. In the current implementation, it doesn't look like we handle draining the send channel once the executor enters the `Draining` state.

---

*[src/lib.rs, line 26 at r3](https://reviewable.io/reviews/onesignal/hyper-client-pool/2#-LCZxCHc5WHYoxHjLZdR:-LCZxCHc5WHYoxHjLZdS:b-8mhknl) ([raw file](https://github.com/onesignal/hyper-client-pool/blob/ce241e14d1bfbd5552429bc3383d4067ba17f1fb/src/lib.rs#L26)):*
> ```Rust
> 
> #[cfg(test)]
> mod tests {
> ```

Should these be made into integration tests that live in `tests/lib.rs`?

---

*[src/lib.rs, line 156 at r3](https://reviewable.io/reviews/onesignal/hyper-client-pool/2#-LCZxPhkAg0Ul4LhaaHu:-LCZxPhkAg0Ul4LhaaHv:b-gqr8x5) ([raw file](https://github.com/onesignal/hyper-client-pool/blob/ce241e14d1bfbd5552429bc3383d4067ba17f1fb/src/lib.rs#L156)):*
> ```Rust
>     }
> 
>     const ONE_SIGNAL_IP_ADDRESSES : [&'static str; 5] = [
> ```

The addresses you have listed here are actually CloudFlare IP addresses. Rather than have a hardcoded list of IPs, we can use CloudFlare's list of IP ranges to check whether `lsof` contains anything matching those IPs. I actually wrote some code to do this last night as part of debugging a potential issue on our loadbalancers. This reads from `stdin` rather than the `lsof` output, but it should be easy to adapt.

```rust
extern crate ipnet;

use std::io::{self, Read, BufRead};
use std::net::IpAddr;

use ipnet::{Contains, IpNet};

static CLOUDFLARE_NETS: &[&str] = &[
    "103.21.244.0/22",
    "103.22.200.0/22",
    "103.31.4.0/22",
    "104.16.0.0/12",
    "108.162.192.0/18",
    "131.0.72.0/22",
    "141.101.64.0/18",
    "162.158.0.0/15",
    "172.64.0.0/13",
    "173.245.48.0/20",
    "188.114.96.0/20",
    "190.93.240.0/20",
    "197.234.240.0/22",
    "198.41.128.0/17",
];

fn run() -> Result<(), Box<::std::error::Error>> {
    let cloudflare_nets = CLOUDFLARE_NETS.iter()
        .map(|net| net.parse::<IpNet>())
        .collect::<Result<Vec<IpNet>, _>>()?;

    let stdin = io::stdin();
    let mut stdin = stdin.lock();

    for line in stdin.lines() {
        let line = line?;
        match line.parse::<IpAddr>() {
            Ok(addr) => {
                let is_cf = cloudflare_nets
                    .iter()
                    .any(|net| net.contains(&addr));

                if !is_cf {
                    println!("{}", line);
                }
            },
            _ => {
                eprintln!("Failed to parse {:?} as ip", line);
            },
        }
    }


    Ok(())
}

fn main() {
    run().unwrap();
}
```

---

*[src/pool.rs, line 60 at r3](https://reviewable.io/reviews/onesignal/hyper-client-pool/2#-LCZz4aw5Wvxqp5cBPzZ:-LCZz4aw5Wvxqp5cBPz_:b-xgzpzf) ([raw file](https://github.com/onesignal/hyper-client-pool/blob/ce241e14d1bfbd5552429bc3383d4067ba17f1fb/src/pool.rs#L60)):*
> ```Rust
>             match handle.send(transaction) {
>                 Err(RequestError::Full(transaction)) => {
>                     ActResult::ValidWithError(Error::Full(transaction))
> ```

Please document the conditions that would lead to these `ActResult` or `RequestError` variants. Given that this is a new lib, readers are unlikely to be familiar with it, and this code should attempt to explain what's going on in each of these cases.

---

*[src/transaction.rs, line 58 at r3](https://reviewable.io/reviews/onesignal/hyper-client-pool/2#-LC_3r2J-DMbtliXUQYc:-LC_3r2J-DMbtliXUQYd:b-adjwm8) ([raw file](https://github.com/onesignal/hyper-client-pool/blob/ce241e14d1bfbd5552429bc3383d4067ba17f1fb/src/transaction.rs#L58)):*
> ```Rust
> }
> 
> struct SpawnedTransaction<D: Deliverable, W: Future, R: Future>
> ```

Bounds are not necessary here

---

*[src/transaction.rs, line 69 at r3](https://reviewable.io/reviews/onesignal/hyper-client-pool/2#-LC_3sYT2Hq_f0U2vKhM:-LC_3sYV2xXPdFBPFfY6:b-adjwm8) ([raw file](https://github.com/onesignal/hyper-client-pool/blob/ce241e14d1bfbd5552429bc3383d4067ba17f1fb/src/transaction.rs#L69)):*
> ```Rust
> }
> 
> impl<D: Deliverable, W: Future, R: Future> Drop for SpawnedTransaction<D, W, R> {
> ```

Bounds are not necessary here

---

*[src/transaction.rs, line 84 at r3](https://reviewable.io/reviews/onesignal/hyper-client-pool/2#-LC_2Z077ALIulygv0_o:-LC_2Z077ALIulygv0_p:b-qdggq7) ([raw file](https://github.com/onesignal/hyper-client-pool/blob/ce241e14d1bfbd5552429bc3383d4067ba17f1fb/src/transaction.rs#L84)):*
> ```Rust
>     D: Deliverable,
>     W: Future<
>         Item=Either<((Response, Vec<u8>), Timeout), ((), R)>,
> ```

I believe we can remove the `PhantomData` since `W` is only constrained in a `where` clause. We would need to remove the generic bound of `R` from the `SpawnedTransaction` type as well.

---

*[src/transaction.rs, line 92 at r3](https://reviewable.io/reviews/onesignal/hyper-client-pool/2#-LC_5PhB5r6f7RMKqxwk:-LC_5PhB5r6f7RMKqxwl:b9v82fe) ([raw file](https://github.com/onesignal/hyper-client-pool/blob/ce241e14d1bfbd5552429bc3383d4067ba17f1fb/src/transaction.rs#L92)):*
> ```Rust
>     >,
> {
>     type Item = ();
> ```

It's surprising to me that this is not modeled as a `Future<Item=Response+Body, Error=TransactionError>`.

---

*[src/transaction.rs, line 146 at r3](https://reviewable.io/reviews/onesignal/hyper-client-pool/2#-LC_4RWq9Mgy2qfYWjx1:-LC_4RWq9Mgy2qfYWjx2:b-c5cwh3) ([raw file](https://github.com/onesignal/hyper-client-pool/blob/ce241e14d1bfbd5552429bc3383d4067ba17f1fb/src/transaction.rs#L146)):*
> ```Rust
>             .map(|deliverable| {
>                 deliverable.complete(delivery_result);
>                 self.task.notify();
> ```

What is the point of notifying the task here? Please add a comment. Maybe also a doc comment where the `task` field is defined.

---

*[src/transaction.rs, line 164 at r3](https://reviewable.io/reviews/onesignal/hyper-client-pool/2#-LC_0VHsD2qZsm6XFRim:-LC_0VHsD2qZsm6XFRin:b-5qnxce) ([raw file](https://github.com/onesignal/hyper-client-pool/blob/ce241e14d1bfbd5552429bc3383d4067ba17f1fb/src/transaction.rs#L164)):*
> ```Rust
>     }
> 
>     pub(crate) fn spawn_request(self, client: &Client<HttpsConnector<HttpConnector>>, handle: &Handle, timeout: Duration, counter: Counter) {
> ```

Break long line

---

*[src/transaction.rs, line 172 at r3](https://reviewable.io/reviews/onesignal/hyper-client-pool/2#-LC_6t7P6MZF1onRlmFZ:-LC_6t7P6MZF1onRlmF_:bz5gb3c) ([raw file](https://github.com/onesignal/hyper-client-pool/blob/ce241e14d1bfbd5552429bc3383d4067ba17f1fb/src/transaction.rs#L172)):*
> ```Rust
>             .and_then(|response| {
>                 let status = response.status();
>                 let headers = response.headers().clone();
> ```

Is there any way to avoid the header clone?

---

*[src/transaction.rs, line 186 at r3](https://reviewable.io/reviews/onesignal/hyper-client-pool/2#-LC_1A3DEk3WbF4Bnm-i:-LC_1A3DEk3WbF4Bnm-j:b-fgcc1z) ([raw file](https://github.com/onesignal/hyper-client-pool/blob/ce241e14d1bfbd5552429bc3383d4067ba17f1fb/src/transaction.rs#L186)):*
> ```Rust
>         match Timeout::new(timeout, handle) {
>             Err(error) => {
>                 deliverable.complete(DeliveryResult::TimeoutError {
> ```

Perhaps `TimeoutError` should be rolled into a `SpawnRequestError`?

---

*[src/transaction.rs, line 208 at r3](https://reviewable.io/reviews/onesignal/hyper-client-pool/2#-LC_4sYZ6IjTY8-RWH_p:-LC_4sYZ6IjTY8-RWH_q:bxsis1g) ([raw file](https://github.com/onesignal/hyper-client-pool/blob/ce241e14d1bfbd5552429bc3383d4067ba17f1fb/src/transaction.rs#L208)):*
> ```Rust
> 
> #[cfg(test)]
> mod tests {
> ```

Please add a test for transaction timeouts. This could be achieved by starting a local server which doesn't send a response.

---


*Comments from [Reviewable](https://reviewable.io/reviews/onesignal/hyper-client-pool/2)*
<!-- Sent from Reviewable.io -->

jwilm commented:
Reviewed 4 of 5 files at r4.
Review status: all files reviewed at latest revision, 21 unresolved discussions.

---

*[src/executor.rs, line 17 at r3](https://reviewable.io/reviews/onesignal/hyper-client-pool/2#-LC_8Rz931iWUUfj9NJZ:-LC_Ca0eEOdBW03D4pTH:bktwvxl) ([raw file](https://github.com/onesignal/hyper-client-pool/blob/ce241e14d1bfbd5552429bc3383d4067ba17f1fb/src/executor.rs#L17)):*
<details><summary><i>Previously, jwilm (Joe Wilm) wrote…</i></summary><blockquote>

Reviewable says `counter` module was removed in latest revision...
</blockquote></details>

You pushed a fix for this during my review :P

---

*[src/lib.rs, line 156 at r3](https://reviewable.io/reviews/onesignal/hyper-client-pool/2#-LCZxPhkAg0Ul4LhaaHu:-LC_Cn41Dpw4L5TT0xmd:b-fi89ev) ([raw file](https://github.com/onesignal/hyper-client-pool/blob/ce241e14d1bfbd5552429bc3383d4067ba17f1fb/src/lib.rs#L156)):*
<details><summary><i>Previously, jwilm (Joe Wilm) wrote…</i></summary><blockquote>

The addresses you have listed here are actually CloudFlare IP addresses. Rather than have a hardcoded list of IPs, we can use CloudFlare's list of IP ranges to check whether `lsof` contains anything matching those IPs. I actually wrote some code to do this last night as part of debugging a potential issue on our loadbalancers. This reads from `stdin` rather than the `lsof` output, but it should be easy to adapt.

```rust
extern crate ipnet;

use std::io::{self, Read, BufRead};
use std::net::IpAddr;

use ipnet::{Contains, IpNet};

static CLOUDFLARE_NETS: &[&str] = &[
    "103.21.244.0/22",
    "103.22.200.0/22",
    "103.31.4.0/22",
    "104.16.0.0/12",
    "108.162.192.0/18",
    "131.0.72.0/22",
    "141.101.64.0/18",
    "162.158.0.0/15",
    "172.64.0.0/13",
    "173.245.48.0/20",
    "188.114.96.0/20",
    "190.93.240.0/20",
    "197.234.240.0/22",
    "198.41.128.0/17",
];

fn run() -> Result<(), Box<::std::error::Error>> {
    let cloudflare_nets = CLOUDFLARE_NETS.iter()
        .map(|net| net.parse::<IpNet>())
        .collect::<Result<Vec<IpNet>, _>>()?;

    let stdin = io::stdin();
    let mut stdin = stdin.lock();

    for line in stdin.lines() {
        let line = line?;
        match line.parse::<IpAddr>() {
            Ok(addr) => {
                let is_cf = cloudflare_nets
                    .iter()
                    .any(|net| net.contains(&addr));

                if !is_cf {
                    println!("{}", line);
                }
            },
            _ => {
                eprintln!("Failed to parse {:?} as ip", line);
            },
        }
    }


    Ok(())
}

fn main() {
    run().unwrap();
}
```
</blockquote></details>

Looks like you just added IPv6 addrs as well. The list of all addresses can be found at https://www.cloudflare.com/ips/

---


*Comments from [Reviewable](https://reviewable.io/reviews/onesignal/hyper-client-pool/2)*
<!-- Sent from Reviewable.io -->

jwilm commented:
Reviewed 2 of 2 files at r5.
Review status: all files reviewed at latest revision, 21 unresolved discussions, some commit checks failed.

---

*[Cargo.toml, line 23 at r5](https://reviewable.io/reviews/onesignal/hyper-client-pool/2#-LC_Y0L7-IEnBo2ZFouj:-LC_Y0L7-IEnBo2ZFouk:b-g94uh7) ([raw file](https://github.com/onesignal/hyper-client-pool/blob/ae75281c197573e9bbfd6b6ed079443be8b855f3/Cargo.toml#L23)):*
> ```TOML
> 
> [replace]
> "tokio-core:0.1.17" = { git = "https://github.com/DarrenTsung/tokio-core", branch = "connection-reuse" }
> ```

What is this about?

---


*Comments from [Reviewable](https://reviewable.io/reviews/onesignal/hyper-client-pool/2)*
<!-- Sent from Reviewable.io -->

DarrenTsung commented:
Review status: 7 of 9 files reviewed at latest revision, 22 unresolved discussions.

---

*[Cargo.toml, line 23 at r5](https://reviewable.io/reviews/onesignal/hyper-client-pool/2#-LC_Y0L7-IEnBo2ZFouj:-LC_iJcvC9bocHY4qXi3:by480xl) ([raw file](https://github.com/onesignal/hyper-client-pool/blob/ae75281c197573e9bbfd6b6ed079443be8b855f3/Cargo.toml#L23)):*
<details><summary><i>Previously, jwilm (Joe Wilm) wrote…</i></summary><blockquote>

What is this about?
</blockquote></details>

You can ignore this, I was investigating connection-reuse not seeming to work on the test server, but actually realized that it was because the `loader` binary was adding new connections everytime. Oops :P.

---

*[src/executor.rs, line 89 at r3](https://reviewable.io/reviews/onesignal/hyper-client-pool/2#-LC_A1Ee2C86pKGSAdxz:-LC_OLYP0ASfwqa4hpVl:b-auxtw5) ([raw file](https://github.com/onesignal/hyper-client-pool/blob/ce241e14d1bfbd5552429bc3383d4067ba17f1fb/src/executor.rs#L89)):*
<details><summary><i>Previously, jwilm (Joe Wilm) wrote…</i></summary><blockquote>

I find this whole `TlsConnector` thing pretty strange. We can't establish our own sockets?

How is hyper doing DNS? We **must not** use the built-in glibc `getaddrinfo` that it previously used via `ToSocketAddrs`. This is why we switched to c-ares at some point. We should switch the resolver to TrustDNS resolver if it's not already using that.

I suppose we could actually implement our own `TlsConnector` and not have to fork Hyper :)
</blockquote></details>

Ah okay, I followed the path down tokio-tls and it does resolve to net::ToSocketAddrs implementation. I will take a look at implementing our own TlsConnector with a 

---


*Comments from [Reviewable](https://reviewable.io/reviews/onesignal/hyper-client-pool/2)*
<!-- Sent from Reviewable.io -->

DarrenTsung commented:
Review status: 7 of 9 files reviewed at latest revision, 22 unresolved discussions.

---

*[src/executor.rs, line 89 at r3](https://reviewable.io/reviews/onesignal/hyper-client-pool/2#-LC_A1Ee2C86pKGSAdxz:-LC_qKICCU86rqYRA9D-:bwcd9fz) ([raw file](https://github.com/onesignal/hyper-client-pool/blob/ce241e14d1bfbd5552429bc3383d4067ba17f1fb/src/executor.rs#L89)):*
<details><summary><i>Previously, DarrenTsung (Darren Tsung) wrote…</i></summary><blockquote>

Ah okay, I followed the path down tokio-tls and it does resolve to net::ToSocketAddrs implementation. I will take a look at implementing our own TlsConnector with a 
</blockquote></details>

Oops, I didn't realize I published this. I opened an issue on hyper to see what their input is: https://github.com/hyperium/hyper/issues/1517

---


*Comments from [Reviewable](https://reviewable.io/reviews/onesignal/hyper-client-pool/2)*
<!-- Sent from Reviewable.io -->

DarrenTsung commented:
Review status: 1 of 9 files reviewed at latest revision, 22 unresolved discussions.

---

*[src/config.rs, line 18 at r3](https://reviewable.io/reviews/onesignal/hyper-client-pool/2#-LCZwl0P0W9YGTPsrbli:-LC_qoFP3C_XNB8NXa6m:b-w3bcuq) ([raw file](https://github.com/onesignal/hyper-client-pool/blob/ce241e14d1bfbd5552429bc3383d4067ba17f1fb/src/config.rs#L18)):*
<details><summary><i>Previously, jwilm (Joe Wilm) wrote…</i></summary><blockquote>

Does hyper 0.11 still handle DNS? Seems like we should be able to plugin a tokio/futures-compatible resolver library and eliminate this option.
</blockquote></details>

Yeah, hyper still handles DNS, see other comment. Will leave this open so we can remove this option.

---

*[src/error.rs, line 9 at r3](https://reviewable.io/reviews/onesignal/hyper-client-pool/2#-LCZy8JoA6GqlYVPjwBW:-LC_rhAA77GOepha8YWA:b-fsydok) ([raw file](https://github.com/onesignal/hyper-client-pool/blob/ce241e14d1bfbd5552429bc3383d4067ba17f1fb/src/error.rs#L9)):*
<details><summary><i>Previously, jwilm (Joe Wilm) wrote…</i></summary><blockquote>

The variants here need some explanation in doc comments. It would also be nice to expose a library `Error` type that buckets these into logical categories or something that is suggestive about how a consumer of this API should handle the errors. 
</blockquote></details>

Refactored the error handling and added doc comments!

There are two types of errors that I wanted to expose because only a subset of errors can happen during `new()`, and I do think that logically you could group error handling between critical errors during Spawn vs the pool being full.

---

*[src/error.rs, line 12 at r3](https://reviewable.io/reviews/onesignal/hyper-client-pool/2#-LCZy0ba-p3pepCmDKDm:-LC_rUpn09lurikNtSuJ:bghws6k) ([raw file](https://github.com/onesignal/hyper-client-pool/blob/ce241e14d1bfbd5552429bc3383d4067ba17f1fb/src/error.rs#L12)):*
<details><summary><i>Previously, jwilm (Joe Wilm) wrote…</i></summary><blockquote>

`Full`? What does that mean?
</blockquote></details>

Renamed to `PoolFull` and added a doc comment. It was previously named `Full`, but I agree that it's not exactly clear. 

---

*[src/error.rs, line 33 at r3](https://reviewable.io/reviews/onesignal/hyper-client-pool/2#-LCZygOc7bXbyMV44ibC:-LCeynAg3TN2aMddYkrA:b-ibbrwx) ([raw file](https://github.com/onesignal/hyper-client-pool/blob/ce241e14d1bfbd5552429bc3383d4067ba17f1fb/src/error.rs#L33)):*
<details><summary><i>Previously, jwilm (Joe Wilm) wrote…</i></summary><blockquote>

This doesn't seem unreachable
</blockquote></details>

Removed with error refactor

---

*[src/error.rs, line 43 at r3](https://reviewable.io/reviews/onesignal/hyper-client-pool/2#-LCZyeVVDBMdhej5jv8k:-LCeypUN1k_uW5p4nv-Z:b-ibbrwx) ([raw file](https://github.com/onesignal/hyper-client-pool/blob/ce241e14d1bfbd5552429bc3383d4067ba17f1fb/src/error.rs#L43)):*
<details><summary><i>Previously, jwilm (Joe Wilm) wrote…</i></summary><blockquote>

Nor does this
</blockquote></details>

Removed with error refactor

---

*[src/executor.rs, line 17 at r3](https://reviewable.io/reviews/onesignal/hyper-client-pool/2#-LC_8Rz931iWUUfj9NJZ:-LCeyqD5ArdM2GId0B97:b32viz) ([raw file](https://github.com/onesignal/hyper-client-pool/blob/ce241e14d1bfbd5552429bc3383d4067ba17f1fb/src/executor.rs#L17)):*
<details><summary><i>Previously, jwilm (Joe Wilm) wrote…</i></summary><blockquote>

You pushed a fix for this during my review :P
</blockquote></details>

:P

---

*[src/executor.rs, line 40 at r3](https://reviewable.io/reviews/onesignal/hyper-client-pool/2#-LC_8Yzc3A3CL6SnFFnh:-LCf0Wzl1E-LDHpmdPYd:bvpwlue) ([raw file](https://github.com/onesignal/hyper-client-pool/blob/ce241e14d1bfbd5552429bc3383d4067ba17f1fb/src/executor.rs#L40)):*
<details><summary><i>Previously, jwilm (Joe Wilm) wrote…</i></summary><blockquote>

We should use a bounded channel here. The `Sender` for a bounded channel includes a `try_send` method which allows attempting a non-blocking send.
</blockquote></details>

UnboundedSender shouldn't block though, right? I mean, it can't block waiting for room as it's unbounded, so it should always be placing a message in the queue.

We handle backpressure over the entire system with the `transaction_counter` as well.

---

*[src/executor.rs, line 159 at r3](https://reviewable.io/reviews/onesignal/hyper-client-pool/2#-LC_C2yZ7ll7JscV9HHM:-LCf0eSmAJ3nIzYWUUb8:bc0i1dk) ([raw file](https://github.com/onesignal/hyper-client-pool/blob/ce241e14d1bfbd5552429bc3383d4067ba17f1fb/src/executor.rs#L159)):*
<details><summary><i>Previously, jwilm (Joe Wilm) wrote…</i></summary><blockquote>

I'm not a big fan of the `Shutdown` message for terminating an executor. Dropping all of the send handles guarantees no messages arrive after detecting the channel has closed. In the current implementation, it doesn't look like we handle draining the send channel once the executor enters the `Draining` state.
</blockquote></details>

Yeah I don't know why I didn't use the pattern of dropping the handle as well. I put a note down that sending a shutdown would drop the handle, but I guess that doesn't account for users that cloned the Handle (if that was possible).

I switched it over to dropping the handle. Thanks!

---

*[src/lib.rs, line 26 at r3](https://reviewable.io/reviews/onesignal/hyper-client-pool/2#-LCZxCHc5WHYoxHjLZdR:-LCf3UkO6JIVzx1kYUg9:b-vgonom) ([raw file](https://github.com/onesignal/hyper-client-pool/blob/ce241e14d1bfbd5552429bc3383d4067ba17f1fb/src/lib.rs#L26)):*
<details><summary><i>Previously, jwilm (Joe Wilm) wrote…</i></summary><blockquote>

Should these be made into integration tests that live in `tests/lib.rs`?
</blockquote></details>

Good point, moved over!

---

*[src/lib.rs, line 156 at r3](https://reviewable.io/reviews/onesignal/hyper-client-pool/2#-LCZxPhkAg0Ul4LhaaHu:-LCfMpzD8a1eb2GoVrDe:b-l41nk2) ([raw file](https://github.com/onesignal/hyper-client-pool/blob/ce241e14d1bfbd5552429bc3383d4067ba17f1fb/src/lib.rs#L156)):*
<details><summary><i>Previously, jwilm (Joe Wilm) wrote…</i></summary><blockquote>

Looks like you just added IPv6 addrs as well. The list of all addresses can be found at https://www.cloudflare.com/ips/
</blockquote></details>

Implemented this, thanks for the reference!

---

*[src/pool.rs, line 60 at r3](https://reviewable.io/reviews/onesignal/hyper-client-pool/2#-LCZz4aw5Wvxqp5cBPzZ:-LCfMwnKA-8eKa6CMEnP:b-fxzi44) ([raw file](https://github.com/onesignal/hyper-client-pool/blob/ce241e14d1bfbd5552429bc3383d4067ba17f1fb/src/pool.rs#L60)):*
<details><summary><i>Previously, jwilm (Joe Wilm) wrote…</i></summary><blockquote>

Please document the conditions that would lead to these `ActResult` or `RequestError` variants. Given that this is a new lib, readers are unlikely to be familiar with it, and this code should attempt to explain what's going on in each of these cases.
</blockquote></details>

Refactored it with new fpool implementation, there are comments when we invalidate the thread because it failed to send now

---

*[src/transaction.rs, line 58 at r3](https://reviewable.io/reviews/onesignal/hyper-client-pool/2#-LC_3r2J-DMbtliXUQYc:-LCfO0Nv85SW_jMFzCnR:b14191z) ([raw file](https://github.com/onesignal/hyper-client-pool/blob/ce241e14d1bfbd5552429bc3383d4067ba17f1fb/src/transaction.rs#L58)):*
<details><summary><i>Previously, jwilm (Joe Wilm) wrote…</i></summary><blockquote>

Bounds are not necessary here
</blockquote></details>

Removed R, thanks!

---

*[src/transaction.rs, line 69 at r3](https://reviewable.io/reviews/onesignal/hyper-client-pool/2#-LC_3sYT2Hq_f0U2vKhM:-LCfO-lLAMYfrJSA0Yvs:b-v6s1yr) ([raw file](https://github.com/onesignal/hyper-client-pool/blob/ce241e14d1bfbd5552429bc3383d4067ba17f1fb/src/transaction.rs#L69)):*
<details><summary><i>Previously, jwilm (Joe Wilm) wrote…</i></summary><blockquote>

Bounds are not necessary here
</blockquote></details>

Removed R

---

*[src/transaction.rs, line 84 at r3](https://reviewable.io/reviews/onesignal/hyper-client-pool/2#-LC_2Z077ALIulygv0_o:-LCfO1bmA33EejWzMVmg:b8xcrxx) ([raw file](https://github.com/onesignal/hyper-client-pool/blob/ce241e14d1bfbd5552429bc3383d4067ba17f1fb/src/transaction.rs#L84)):*
<details><summary><i>Previously, jwilm (Joe Wilm) wrote…</i></summary><blockquote>

I believe we can remove the `PhantomData` since `W` is only constrained in a `where` clause. We would need to remove the generic bound of `R` from the `SpawnedTransaction` type as well.
</blockquote></details>

Removed R + PhantomData, thanks!

---

*[src/transaction.rs, line 92 at r3](https://reviewable.io/reviews/onesignal/hyper-client-pool/2#-LC_5PhB5r6f7RMKqxwk:-LCfOOqj-D9C3litU1qW:b92c589) ([raw file](https://github.com/onesignal/hyper-client-pool/blob/ce241e14d1bfbd5552429bc3383d4067ba17f1fb/src/transaction.rs#L92)):*
<details><summary><i>Previously, jwilm (Joe Wilm) wrote…</i></summary><blockquote>

It's surprising to me that this is not modeled as a `Future<Item=Response+Body, Error=TransactionError>`.
</blockquote></details>

That would work - as well as an `and_then()` that reports the result to the dispatcher(?), right? But I'm not sure how that would work with the `Drop` handling since we need to add the impl `Drop` for some type.

---

*[src/transaction.rs, line 146 at r3](https://reviewable.io/reviews/onesignal/hyper-client-pool/2#-LC_4RWq9Mgy2qfYWjx1:-LCfPa6R9iCGoT8EAOOp:b-csoc7) ([raw file](https://github.com/onesignal/hyper-client-pool/blob/ce241e14d1bfbd5552429bc3383d4067ba17f1fb/src/transaction.rs#L146)):*
<details><summary><i>Previously, jwilm (Joe Wilm) wrote…</i></summary><blockquote>

What is the point of notifying the task here? Please add a comment. Maybe also a doc comment where the `task` field is defined.
</blockquote></details>

Added comments + a doc comment. Basically the executor is waiting for the transactions to finish, but is never woken up if not notified by transactions finishing.

---

*[src/transaction.rs, line 164 at r3](https://reviewable.io/reviews/onesignal/hyper-client-pool/2#-LC_0VHsD2qZsm6XFRim:-LCfQ3YK64a84PUJi0H4:b-896fja) ([raw file](https://github.com/onesignal/hyper-client-pool/blob/ce241e14d1bfbd5552429bc3383d4067ba17f1fb/src/transaction.rs#L164)):*
<details><summary><i>Previously, jwilm (Joe Wilm) wrote…</i></summary><blockquote>

Break long line
</blockquote></details>

Done!

---

*[src/transaction.rs, line 172 at r3](https://reviewable.io/reviews/onesignal/hyper-client-pool/2#-LC_6t7P6MZF1onRlmFZ:-LCfQ4fg4WkrswMER0rl:b-v1rcvs) ([raw file](https://github.com/onesignal/hyper-client-pool/blob/ce241e14d1bfbd5552429bc3383d4067ba17f1fb/src/transaction.rs#L172)):*
<details><summary><i>Previously, jwilm (Joe Wilm) wrote…</i></summary><blockquote>

Is there any way to avoid the header clone?
</blockquote></details>

Unfortunately not if we want to return the Response as part of the DeliveryResult.

Spoke to seanmonstar about this and possibly better handling in hyper 0.12.

---

*[src/transaction.rs, line 186 at r3](https://reviewable.io/reviews/onesignal/hyper-client-pool/2#-LC_1A3DEk3WbF4Bnm-i:-LCfQQs-C7JryF4r2_vi:b-epjoep) ([raw file](https://github.com/onesignal/hyper-client-pool/blob/ce241e14d1bfbd5552429bc3383d4067ba17f1fb/src/transaction.rs#L186)):*
<details><summary><i>Previously, jwilm (Joe Wilm) wrote…</i></summary><blockquote>

Perhaps `TimeoutError` should be rolled into a `SpawnRequestError`?
</blockquote></details>

I think the naming of the errors might be confusing, there's:
SpawnError -> returned if constructing a new client for the Pool fails
Error -> contains SpawnError + PoolFull, all error cases when sending a new request.

Then there's also the types in `DeliveryResult`, this is returning an `TimeoutError` which is when the `Timeout` type errored. There's also `DeliveryResult::Timeout` which is when the connection timed out. 

---

*[src/transaction.rs, line 208 at r3](https://reviewable.io/reviews/onesignal/hyper-client-pool/2#-LC_4sYZ6IjTY8-RWH_p:-LCfS1mpA1EHg--sWgmm:b-xm783k) ([raw file](https://github.com/onesignal/hyper-client-pool/blob/ce241e14d1bfbd5552429bc3383d4067ba17f1fb/src/transaction.rs#L208)):*
<details><summary><i>Previously, jwilm (Joe Wilm) wrote…</i></summary><blockquote>

Please add a test for transaction timeouts. This could be achieved by starting a local server which doesn't send a response.
</blockquote></details>

I have a test for timeout in the integration tests, see `timeout_works_as_expected()`

---


*Comments from [Reviewable](https://reviewable.io/reviews/onesignal/hyper-client-pool/2)*
<!-- Sent from Reviewable.io -->

DarrenTsung commented:
Review status: 1 of 10 files reviewed at latest revision, 17 unresolved discussions.

---

*[src/config.rs, line 18 at r3](https://reviewable.io/reviews/onesignal/hyper-client-pool/2#-LCZwl0P0W9YGTPsrbli:-LCkI8wU1YvwGSJ-YTu5:b-mvw84d) ([raw file](https://github.com/onesignal/hyper-client-pool/blob/ce241e14d1bfbd5552429bc3383d4067ba17f1fb/src/config.rs#L18)):*
<details><summary><i>Previously, DarrenTsung (Darren Tsung) wrote…</i></summary><blockquote>

Yeah, hyper still handles DNS, see other comment. Will leave this open so we can remove this option.
</blockquote></details>

Switched to a custom HttpConnector that uses c-ares DNS resolver. I tried with trust-dns first, but load testing revealed panics in the resolution flow.

---

*[src/executor.rs, line 89 at r3](https://reviewable.io/reviews/onesignal/hyper-client-pool/2#-LC_A1Ee2C86pKGSAdxz:-LCkI_9z2fZCU92BO9wH:b-mvw84d) ([raw file](https://github.com/onesignal/hyper-client-pool/blob/ce241e14d1bfbd5552429bc3383d4067ba17f1fb/src/executor.rs#L89)):*
<details><summary><i>Previously, DarrenTsung (Darren Tsung) wrote…</i></summary><blockquote>

Oops, I didn't realize I published this. I opened an issue on hyper to see what their input is: https://github.com/hyperium/hyper/issues/1517
</blockquote></details>

Switched to a custom HttpConnector that uses c-ares DNS resolver. I tried with trust-dns first, but load testing revealed panics in the resolution flow.

---


*Comments from [Reviewable](https://reviewable.io/reviews/onesignal/hyper-client-pool/2)*
<!-- Sent from Reviewable.io -->

DarrenTsung commented:
Review status: 9 of 10 files reviewed at latest revision, 7 unresolved discussions.

---

*[src/config.rs, line 18 at r3](https://reviewable.io/reviews/onesignal/hyper-client-pool/2#-LCZwl0P0W9YGTPsrbli:-LDJTZCL2u7kOzqqy5xR:brpu2cd) ([raw file](https://github.com/onesignal/hyper-client-pool/blob/ce241e14d1bfbd5552429bc3383d4067ba17f1fb/src/config.rs#L18)):*
<details><summary><i>Previously, jwilm (Joe Wilm) wrote…</i></summary><blockquote>

We should work with the TrustDNS project to resolve whatever issues we are having.
</blockquote></details>

Figured out that trying to run a ResolverFuture with tokio-core was causing the issue, moved over to `trust-dns`! 

---

*[src/error.rs, line 42 at r7](https://reviewable.io/reviews/onesignal/hyper-client-pool/2#-LCfWyzC6jxWQGiLED_C:-LDJTrOy4Yc9HIouCP6Q:bb6un1j) ([raw file](https://github.com/onesignal/hyper-client-pool/blob/62c16f94b28b86ff1fca2d1cdf5f4b995f2894ce/src/error.rs#L42)):*
<details><summary><i>Previously, jwilm (Joe Wilm) wrote…</i></summary><blockquote>

In the standard library, the naming convention for this type of method is `into_inner`.
</blockquote></details>

Done

---

*[src/executor.rs, line 40 at r3](https://reviewable.io/reviews/onesignal/hyper-client-pool/2#-LC_8Yzc3A3CL6SnFFnh:-LDJZRFV04U7isMkTHkP:b1mokxu) ([raw file](https://github.com/onesignal/hyper-client-pool/blob/ce241e14d1bfbd5552429bc3383d4067ba17f1fb/src/executor.rs#L40)):*
<details><summary><i>Previously, jwilm (Joe Wilm) wrote…</i></summary><blockquote>

It's more about knowing worst-case how much memory this channel will use.
</blockquote></details>

Right, good point. Actually in this case this channel is completely bounded by `max_transactions` as the Counter is cloned and sent as part of the message.

Along with that, the executor thread would immediately spawn the transaction as part of consuming it while it has the receiver. No chance it would be backlogged unless the receiver is dropped / thread doesn't get woken up, it just seems weird to add a failure case that shouldn't happen (given the in-variants of tokio / the channel type).

---

*[src/executor.rs, line 65 at r8](https://reviewable.io/reviews/onesignal/hyper-client-pool/2#-LDEXsDq7ooMMPLGCI5v:-LDJYeRHAj4dZjH-sAhL:b-896fja) ([raw file](https://github.com/onesignal/hyper-client-pool/blob/90d2125aa7d628b98be6d32cc24b72434bda032b/src/executor.rs#L65)):*
<details><summary><i>Previously, jwilm (Joe Wilm) wrote…</i></summary><blockquote>

We should prefer a doc comment rather than having unnecessary lines of code.
</blockquote></details>

Done!

---

*[src/pool.rs, line 67 at r8](https://reviewable.io/reviews/onesignal/hyper-client-pool/2#-LDEbGPi9CbO1994_Vnc:-LDJXNMwBuh3Hh6ZL10O:bli3ht9) ([raw file](https://github.com/onesignal/hyper-client-pool/blob/90d2125aa7d628b98be6d32cc24b72434bda032b/src/pool.rs#L67)):*
<details><summary><i>Previously, jwilm (Joe Wilm) wrote…</i></summary><blockquote>

It's a little surprising that the consumer is responsible for this. Maybe we could use an intermediate abstraction on the fpool?
</blockquote></details>

`fpool` doesn't know what the objects are being used for, which is why an outer user must invalidate the object in an error case. 

I guess we could wrap the fpool in some sort of type that handles it, but I'm not sure what this type would be left with? 

---

*[src/transaction.rs, line 92 at r3](https://reviewable.io/reviews/onesignal/hyper-client-pool/2#-LC_5PhB5r6f7RMKqxwk:-LDJUNh-5WmvqbIHeEbJ:b-2zvz4w) ([raw file](https://github.com/onesignal/hyper-client-pool/blob/ce241e14d1bfbd5552429bc3383d4067ba17f1fb/src/transaction.rs#L92)):*
<details><summary><i>Previously, jwilm (Joe Wilm) wrote…</i></summary><blockquote>

How would that change where `Drop` is implemented?
</blockquote></details>

Because `SpawnedTransaction` is the type that ensures that the transaction result is always sent to the deliverable, the `.work` field in `SpawnedTransaction` is the type you're expecting (I think).

The transaction future that returns an Item=Response+Body, Error=Timeout/TransationError is the chained, generalized type `W`.

Basically we have to have some type that has Item=(), Error=() because it needs to also impl `Drop` and notify the deliverable in that case. Even if we had a nice future type which returned `Response+Body` it would have to be wrapped in a struct that implements `Drop`+ has ownership of the Deliverable. Because that struct has ownership of Deliverable, it would also need to be the future that sends the real result to Deliverable as well.

---


*Comments from [Reviewable](https://reviewable.io/reviews/onesignal/hyper-client-pool/2)*
<!-- Sent from Reviewable.io -->

jwilm commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 1 files at r9, 2 of 2 files at r10.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io/reviews/onesignal/hyper-client-pool/2#-:-LDOnUrmD0UML-sFmsQg:bnfp4nl)*
<!-- Sent from Reviewable.io -->
